bigtable/metrics.go

```diff
@@ -0,0 +1,487 @@
+// It is used to store metric instruments, attribute values
+// and other data required to obtain and record them
+type builtinMetricsTracer struct {
+	ctx            context.Context
+	builtInEnabled bool
+	opStartTime    time.Time
+
+	// attributes that are specific to a client instance and
+	// do not change across different function calls on client
+	clientAttributes []attribute.KeyValue
+
+	instrumentOperationLatencies metric.Float64Histogram
+	instrumentServerLatencies    metric.Float64Histogram
+	instrumentAttemptLatencies   metric.Float64Histogram
+	instrumentRetryCount         metric.Int64Counter
+
+	tableName    string
+	appProfileID string
+	method       string
+	isStreaming  bool
+
+	// gRPC status code
+	status string
+
+	// Contains the header response metadata which is used to extract cluster and zone
+	headerMD *metadata.MD
+
+	// Contains the trailer response metadata which is used to extract cluster and zone
+	trailerMD *metadata.MD
+
+	attemptCount int64
+}
+
+func (tf *builtinMetricsTracerFactory) newBuiltinMetricsTracer(ctx context.Context, tableName, appProfile string, isStreaming bool) builtinMetricsTracer {
+	headerMD := metadata.New(nil)
+	trailerMD := metadata.New(nil)
+	return builtinMetricsTracer{
+		ctx:              ctx,
+		builtInEnabled:   tf.builtinEnabled,
+		opStartTime:      time.Now(),
+		clientAttributes: tf.clientAttributes,
+
+		instrumentOperationLatencies: tf.operationLatencies,
+		instrumentServerLatencies:    tf.serverLatencies,
+		instrumentAttemptLatencies:   tf.attemptLatencies,
+		instrumentRetryCount:         tf.retryCount,
+
+		tableName:    tableName,
+		appProfileID: appProfile,
+		isStreaming:  isStreaming,
+		status:       "",
+		headerMD:     &headerMD,
+		trailerMD:    &trailerMD,
+		attemptCount: 0,
+	}
+}
+
+// recordOperationCompletion records as many metrics as it can and does not return error
+func (mt *builtinMetricsTracer) recordOperationCompletion() {
+	if !mt.builtInEnabled {
+		return
+	}
+
+	// Calculate elapsed time
+	elapsedTimeMs := float64(time.Since(mt.opStartTime).Nanoseconds()) / 1000000
+
+	// Attributes for operation_latencies
+	opLatCurrCallAttrs, opLatErr := mt.toOtelMetricAttrs(metricNameOperationLatencies)
+	opLatAllAttrs := metric.WithAttributes(append(mt.clientAttributes, opLatCurrCallAttrs...)...)
+
+	// Attributes for retry_count
+	retryCntCurrCallAttrs, retryCountErr := mt.toOtelMetricAttrs(metricNameRetryCount)
+	retryCntAllAttrs := metric.WithAttributes(append(mt.clientAttributes, retryCntCurrCallAttrs...)...)
+
+	if opLatErr == nil {
+		mt.instrumentOperationLatencies.Record(mt.ctx, elapsedTimeMs, opLatAllAttrs)
+	}
+
+	// Only record when retry count is greater than 0 so the retry
+	// graph will be less confusing
+	if mt.attemptCount > 1 {
+		if retryCountErr == nil {
+			mt.instrumentRetryCount.Add(mt.ctx, mt.attemptCount-1, retryCntAllAttrs)
+		}
+	}
+}
+
+func (mt *builtinMetricsTracer) gaxInvokeWithRecorder(ctx context.Context, method string,
```

Comment:
I think it might be nicer to move the inversion of control outside of this class. This way the client is the primary flow of control and this is just a helper rather then threading control in and out of the client

Reply:
It should also make for easier testing in isolation

Reply:
Done